src/ceph_nfs.py

```diff
@@ -0,0 +1,387 @@
+        self.framework.observe(charm.on["peers"].relation_departed, self._on_ceph_peers)
+        self.framework.observe(charm.on["peers"].relation_changed, self._on_ceph_peers)
+
+    def _on_relation_changed(self, event):
+        """Prepare relation for data from requiring side."""
+        if not self.model.unit.is_leader():
+            return
+
+        logger.info("_on_relation_changed event")
+
+        if not self.charm.ready_for_service():
+            logger.info("Not processing request as service is not yet ready")
+            event.defer()
+            return
+
+        if ceph.get_osd_count() == 0:
+            logger.info("Storage not available, deferring event.")
+            event.defer()
+            return
+
+        self.on.ceph_nfs_connected.emit(event.relation)
+
+    def _on_relation_departed(self, event):
+        """Cleanup relation after departure."""
+        if not self.model.unit.is_leader() or event.relation.app == self.charm.app:
+            return
+
+        logger.info("_on_relation_departed event")
+        self.on.ceph_nfs_departed.emit(event.relation)
+
+    def _on_ceph_peers(self, event):
+        """Handle ceph peers relation events."""
+        if not self.model.unit.is_leader():
+            return
+
+        if ceph.get_osd_count() == 0:
+            logger.info("Storage not available, deferring event.")
+            event.defer()
+            return
+
+        logger.info("_on_ceph_peers event")
+
+        # Mon addrs might have changed, update the relation data.
+        # Additionally, new nodes may have been added, which could be added to
+        # NFS clusters.
+        self.on.ceph_nfs_reconcile.emit(event.relation)
+
+
+class CephNfsProviderHandler(RelationHandler):
+    """Handler for the ceph-nfs relation."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str,
+        callback_f: Callable,
+    ):
+        super().__init__(charm, relation_name, callback_f)
+
+    def setup_event_handler(self) -> Object:
+        """Configure event handlers for an ceph-nfs-client interface."""
+        logger.debug("Setting up ceph-nfs-client event handler")
+
+        ceph_nfs = CephNfsProvides(self.charm, self.relation_name)
+        self.framework.observe(ceph_nfs.on.ceph_nfs_connected, self._on_ceph_nfs_connected)
+        self.framework.observe(ceph_nfs.on.ceph_nfs_reconcile, self._on_ceph_nfs_reconcile)
+        self.framework.observe(ceph_nfs.on.ceph_nfs_departed, self._on_ceph_nfs_departed)
+
+        return ceph_nfs
+
+    @property
+    def ready(self) -> bool:
+        """Check if handler is ready."""
+        return True
+
+    def set_status(self, status: compound_status.Status) -> None:
+        """Set the status based on current state."""
+        relations = self.model.relations[self.relation_name]
+        if not relations:
+            # If there are no ceph-nfs relations, no need to block this.
+            status.set(ActiveStatus(""))
```

Comment:
i think the behavior should be to do nothing while as setting active status may ow any existing status.

Reply:
This will set the relation's status, not the entire's charm status. Other relations have their own statuses, and the charm as well.

Reply:
Got it, I checked with @gboutry and there is a resolver mechanism for this. NEAT!